---
title: "IDO1 gene expression analysis"
output: html_notebook
---

```{r setup, include=FALSE}

library(tidyverse)
library(viridisLite)
library(readxl)
library(ggplot2)
library(cowplot)
library(Seurat)
library(scales)
library(readr)

theme_set(theme_cowplot())

#colors to use
use_colors <- c(
  `Epithelial cells` = "brown2",
  `Ovarian stroma cells` = "#bcbddc",
  `Myeloid cells` = "darkgoldenrod2",
  `T/NK cells` = "#7A0177",
  `B cells` = "#006837",
  `B/Plasma cells` = "#006837",
  Fibroblasts = "#08519C",
  `Smooth muscle cells` = "#6bAEd6",
  `Mesothelial cells` = "#78c679",
  `Endothelial cells` = "#EB96EB",
  `Plasma cells` = "chocolate4",
  Undefined = "grey")

```

```{r, include=FALSE}

olbrecht <- readRDS("objects/olbrecht_integrated_annotated.RDS")

olalekan <- readRDS("objects/olalekan_integrated_annotated.RDS")

```

# Cell types in different datasets {.tabset}

## Olbrecht et al.
```{r, echo=FALSE}

# cell type plots
DimPlot(olbrecht, group.by = "cell_type", cols = use_colors, pt.size = 1)
ggsave("output/olbrecht_dimplot_celltypes.png", height = 5, width = 7)
ggsave("output/olbrecht_dimplot_celltypes.pdf", height = 5, width = 7)

cell_counts <- FetchData(subset(olbrecht, subset = sample_type == "Tumor"), c("cell_type", "patient_id", "sample_id"))

ggplot(cell_counts) +
  geom_bar(aes(x = patient_id, fill = cell_type), position = "fill") +
  scale_fill_manual(values = use_colors)

ggplot(cell_counts) +
  geom_bar(aes(x = patient_id, fill = cell_type)) +
  scale_fill_manual(values = use_colors)

```

## Olalekan et al.
```{r, echo=FALSE}

# cell type plots
DimPlot(olalekan, group.by = "cell_type", cols = use_colors, pt.size = 1)
ggsave("output/olalekan_dimplot_celltypes.png", height = 5, width = 7)
ggsave("output/olalekan_dimplot_celltypes.pdf", height = 5, width = 7)

cell_counts <- FetchData(olalekan, c("cell_type", "sample_id"))

ggplot(cell_counts) +
  geom_bar(aes(x = sample_id, fill = cell_type), position = "fill") +
  scale_fill_manual(values = use_colors)

ggplot(cell_counts) +
  geom_bar(aes(x = sample_id, fill = cell_type)) +
  scale_fill_manual(values = use_colors)

```

# Expression of IDO1 in different datasets {.tabset}

```{r, include=FALSE}
gene <- "IDO1"
DefaultAssay(olbrecht) <- "RNA"
DefaultAssay(olalekan) <- "RNA"
```

## Olbrecht et al.
```{r, echo=FALSE, message=FALSE, warning=FALSE}

# plot IDO1 gene expression
FeaturePlot(olbrecht, gene, order = TRUE, pt.size = 1, cols = c("lightblue", "darkred"))
ggsave("output/olbrecht_featureplot_ido1.png", height = 5, width = 6)
ggsave("output/olbrecht_featureplot_ido1.pdf", height = 5, width = 6)

VlnPlot(olbrecht, gene, group.by = "cell_type", pt.size = 0, cols = use_colors) + NoLegend()

```

```{r, include=FALSE}

# binarize tumor cells in  positive and negative
ovca_obj_epi <- subset(olbrecht, subset = cell_type == "Epithelial cells") %>% ScaleData()
ovca_obj_epi_data <- FetchData(ovca_obj_epi, vars = c(paste0(gene), "patient_id")) %>% mutate(type = ifelse(eval(parse(text=paste0(gene))) > 0, "positive", "negative"))
ovca_obj_epi <- AddMetaData(ovca_obj_epi, metadata = ovca_obj_epi_data)

# cell cycles in IDO1-positive vs. IDO1-negative tumor cells
ovca_obj_epi_cc_data <- FetchData(ovca_obj_epi, c("type", "Phase"))

# differentially expressed genes in IDO1-positive vs. IDO1-negative tumor cells
Idents(ovca_obj_epi) <- ovca_obj_epi@meta.data$type
markers <- FindAllMarkers(ovca_obj_epi, only.pos = T, min.pct = 0.2, min.diff.pct = 0.2)
top_markers <- markers %>% group_by(cluster) %>% top_n(10, wt = avg_log2FC)

```

```{r, echo=FALSE}

# cell cycle phase in IDO1-positive vs. IDO1-negative tumor cells
ggplot(data = ovca_obj_epi_cc_data) + 
  geom_bar(mapping = aes(x = type, fill = Phase, ), position = "fill")

# differentially expressed genes in IDO1-positive vs. IDO1-negative tumor cells
DoHeatmap(ovca_obj_epi, features = top_markers$gene, group.by = "type", draw.lines = F, slot = "counts") + scale_fill_viridis_c()
ggsave("output/olbrecht_heatmap_ido1posneg.pdf")
ggsave("output/olbrecht_heatmap_ido1posneg.png")
write.csv(markers, "output/ido1_deg_olbrecht.csv")

# frequency of IDO1-positive vs. IDO1-negative tumor cells across patients
ggplot(data = ovca_obj_epi_data) + 
  geom_bar(mapping = aes(x = patient_id, fill = type), position = "fill") +
  scale_fill_discrete(type = c("lightblue", "darkred"))
ggsave("output/olbrecht_barplot_ido1posneg.pdf")

```

```{r, include=FALSE}

# add Hallmark signature scores

broad_pws <- read_lines("data/h.all.v6.2.symbols.gmt") %>%
  lapply(str_split, "\\t") %>% 
  unlist(recursive = F) %>% 
  lapply(function(x) setNames(list(x[-c(1:2)]), x[1])) %>% 
  unlist(recursive = F)

ovca_obj_epi <- AddModuleScore(object = ovca_obj_epi, features = broad_pws, name = names(broad_pws))



# add APM signatures scores

APM_signatures <- list(c("B2M", "CALR", "NLRC5", "PSMB9", "PSME1", "PSME3", "RFX5", "HSP90AB1"),
                       c("PSMB5", "PSMB6", "PSMB7", "PSMB8", "PSMB9", "PSMB10", "TAP1", "TAP2", "ERAP1", "ERAP2", "CANX", "CALR", "PDIA3", "TAPBP", "B2M", "HLA-A", "HLA-B", "HLA-C"))
KEGG <- read_tsv("data/KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION.v2022.1.gmt") %>% colnames() 
APM_signatures[[3]] <- KEGG[3:length(KEGG)]
ovca_obj_epi <- AddModuleScore(ovca_obj_epi, APM_signatures, name = c("Thompson", "Chen", "KEGG"))

# plots
VlnPlot(ovca_obj_epi, features = c("HALLMARK_INTERFERON_ALPHA_RESPONSE18",
                                   "HALLMARK_INTERFERON_GAMMA_RESPONSE19", 
                                   "HALLMARK_TGF_BETA_SIGNALING6"), 
        group.by = "type", pt.size = 0, cols = c("lightblue", "darkred"))
ggsave("output/olbrecht_vlnplot_hallmark.pdf")

VlnPlot(ovca_obj_epi, features = c("Thompson1", 
                                   "Chen2", 
                                   "KEGG3"), 
        group.by = "type", pt.size = 0, cols = c("lightblue", "darkred"))
ggsave("output/olbrecht_vlnplot_apm.pdf")

```

### Myeloid cells

```{r, include=FALSE}

# subset myeloid immune cells and re-cluster
DefaultAssay(olbrecht) <- "integrated"
Myeloid <- subset(olbrecht, subset = cell_type == "Myeloid cells")

Myeloid <- FindVariableFeatures(Myeloid)
Myeloid <- ScaleData(Myeloid)
Myeloid <- RunPCA(Myeloid)
ElbowPlot(Myeloid, ndims = 50)
Myeloid <- FindNeighbors(Myeloid, dims = 1:10)
Myeloid <- FindClusters(Myeloid, resolution = c(0.5, 1, 2))
Myeloid <- RunUMAP(Myeloid, dims = 1:10)

DimPlot(Myeloid, group.by = "patient_id")
DimPlot(Myeloid, group.by = "integrated_snn_res.1", label = TRUE)

# use RNA slot for DEG analysis
DefaultAssay(Myeloid) <- "RNA"
Myeloid <- ScaleData(Myeloid)

```

```{r, echo=FALSE}

# myeloid immune cell markers for cell type annotation
myeloid_markers <- c("FCN1", "S100A8", "CD14", "FCGR3A", "CD68", "MARCO", "FABP4", "FCGR1A", "C1QA", "APOC1", "LGMN", "CTSB", "FCER1A", "CD1C", "CLEC9A", "LILRA4", "CLEC4C", "IL3RA", "NRP1", "CLEC10A", "PTCRA", "CCR7", "LAMP3", "CPA3", "KIT", "MKI67", "EPCAM")

DotPlot(Myeloid, features = myeloid_markers, group.by = "integrated_snn_res.1") + coord_flip()

```

```{r, echo=FALSE}

# myeloid immune cell type annotation
annotation_curated_myeloid <- read_excel("curated_annotation/curated_annotation_olbrecht_integrated_Myeloid.xlsx")

Idents(Myeloid) <- Myeloid$integrated_snn_res.1
new_ids_main <- annotation_curated_myeloid$cell_subtype
names(new_ids_main) <- levels(Myeloid)
Myeloid <- RenameIdents(Myeloid, new_ids_main)
Myeloid@meta.data$cell_subtype <- Idents(Myeloid)

DimPlot(Myeloid, group.by = "cell_subtype", label = T)

```

```{r, echo=FALSE}

cell_counts <- FetchData(Myeloid, c("cell_subtype", "patient_id", "sample_id")) %>% filter(cell_subtype != "Contamination")

ggplot(cell_counts) +
  geom_bar(aes(x = patient_id, fill = cell_subtype), position = "fill")

ggplot(cell_counts) +
  geom_bar(aes(x = patient_id, fill = cell_subtype))

FeaturePlot(Myeloid, "IDO1", order = TRUE)

VlnPlot(Myeloid, "IDO1", pt.size = 1)
ggsave("output/olbrecht_vlnplot_myeloid_ido1.pdf", height = 5, width = 7)

```







## Olalekan et al.

```{r, echo=FALSE, message=FALSE, warning=FALSE}

# plot IDO1 gene expression
FeaturePlot(olalekan, gene, order = TRUE, pt.size = 1, cols = c("lightblue", "darkred"))
ggsave("output/olalekan_featureplot_ido1.png", height = 5, width = 6)
ggsave("output/olalekan_featureplot_ido1.pdf", height = 5, width = 6)

VlnPlot(olalekan, gene, group.by = "cell_type", pt.size = 0, cols = use_colors) + NoLegend()

```

```{r, include=FALSE}

# binarize tumor cells in  positive and negative
ovca_obj_epi <- subset(olalekan, subset = cell_type == "Epithelial cells") %>% ScaleData()
ovca_obj_epi_data <- FetchData(ovca_obj_epi, vars = c(paste0(gene), "sample_id")) %>% mutate(type = ifelse(eval(parse(text=paste0(gene))) > 0, "positive", "negative"))
ovca_obj_epi <- AddMetaData(ovca_obj_epi, metadata = ovca_obj_epi_data)

# cell cycles in IDO1-positive vs. IDO1-negative tumor cells
ovca_obj_epi_cc_data <- FetchData(ovca_obj_epi, c("type", "Phase"))

# differentially expressed genes in IDO1-positive vs. IDO1-negative tumor cells
Idents(ovca_obj_epi) <- ovca_obj_epi@meta.data$type
markers <- FindAllMarkers(ovca_obj_epi, only.pos = T, min.pct = 0.2, min.diff.pct = 0.2)
top_markers <- markers %>% group_by(cluster) %>% top_n(10, wt = avg_log2FC)

```

```{r, echo=FALSE}

# cell cycle phase in IDO1-positive vs. IDO1-negative tumor cells
ggplot(data = ovca_obj_epi_cc_data) + 
  geom_bar(mapping = aes(x = type, fill = Phase, ), position = "fill")

# differentially expressed genes in IDO1-positive vs. IDO1-negative tumor cells
DoHeatmap(ovca_obj_epi, features = top_markers$gene, group.by = "type", draw.lines = F, slot = "counts") + scale_fill_viridis_c()
ggsave("output/olalekan_heatmap_ido1posneg.pdf")
ggsave("output/olalekan_heatmap_ido1posneg.png")
write.csv(markers, "output/ido1_deg_olalekan.csv")

# frequency of IDO1-positive vs. IDO1-negative tumor cells across patients
ggplot(data = ovca_obj_epi_data) + 
  geom_bar(mapping = aes(x = sample_id, fill = type), position = "fill") +
  scale_fill_discrete(type = c("lightblue", "darkred"))
ggsave("output/olalekan_barplot_ido1posneg.pdf")

```

```{r, include=FALSE}

# add Hallmark signature scores

broad_pws <- read_lines("data/h.all.v6.2.symbols.gmt") %>%
  lapply(str_split, "\\t") %>% 
  unlist(recursive = F) %>% 
  lapply(function(x) setNames(list(x[-c(1:2)]), x[1])) %>% 
  unlist(recursive = F)

ovca_obj_epi <- AddModuleScore(object = ovca_obj_epi, features = broad_pws, name = names(broad_pws))



# add APM signatures scores

APM_signatures <- list(c("B2M", "CALR", "NLRC5", "PSMB9", "PSME1", "PSME3", "RFX5", "HSP90AB1"),
                       c("PSMB5", "PSMB6", "PSMB7", "PSMB8", "PSMB9", "PSMB10", "TAP1", "TAP2", "ERAP1", "ERAP2", "CANX", "CALR", "PDIA3", "TAPBP", "B2M", "HLA-A", "HLA-B", "HLA-C"))
KEGG <- read_tsv("data/KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION.v2022.1.gmt") %>% colnames() 
APM_signatures[[3]] <- KEGG[3:length(KEGG)]
ovca_obj_epi <- AddModuleScore(ovca_obj_epi, APM_signatures, name = c("Thompson", "Chen", "KEGG"))

# plots
VlnPlot(ovca_obj_epi, features = c("HALLMARK_INTERFERON_ALPHA_RESPONSE18",
                                   "HALLMARK_INTERFERON_GAMMA_RESPONSE19", 
                                   "HALLMARK_TGF_BETA_SIGNALING6"), 
        group.by = "type", pt.size = 0, cols = c("lightblue", "darkred"))
ggsave("output/olalekan_vlnplot_hallmark.pdf")

VlnPlot(ovca_obj_epi, features = c("Thompson1", 
                                   "Chen2", 
                                   "KEGG3"), 
        group.by = "type", pt.size = 0, cols = c("lightblue", "darkred"))
ggsave("output/olalekan_vlnplot_apm.pdf")

```

### Myeloid cells

```{r, include=FALSE}

# subset myeloid immune cells and re-cluster
DefaultAssay(olalekan) <- "integrated"
Myeloid <- subset(olalekan, subset = cell_type == "Myeloid cells")

Myeloid <- FindVariableFeatures(Myeloid)
Myeloid <- ScaleData(Myeloid)
Myeloid <- RunPCA(Myeloid)
ElbowPlot(Myeloid, ndims = 50)
Myeloid <- FindNeighbors(Myeloid, dims = 1:10)
Myeloid <- FindClusters(Myeloid, resolution = c(0.5, 1, 2))
Myeloid <- RunUMAP(Myeloid, dims = 1:10)

DimPlot(Myeloid, group.by = "sample_id")
DimPlot(Myeloid, group.by = "integrated_snn_res.1", label = TRUE)

# use RNA slot for DEG analysis
DefaultAssay(Myeloid) <- "RNA"
Myeloid <- ScaleData(Myeloid)

```

```{r, echo=FALSE}

# myeloid immune cell markers for cell type annotation
myeloid_markers <- c("FCN1", "S100A8", "CD14", "FCGR3A", "CD68", "MARCO", "FABP4", "FCGR1A", "C1QA", "APOC1", "LGMN", "CTSB", "FCER1A", "CD1C", "CLEC9A", "LILRA4", "CLEC4C", "IL3RA", "NRP1", "CLEC10A", "PTCRA", "CCR7", "LAMP3", "CPA3", "KIT", "MKI67", "EPCAM")

DotPlot(Myeloid, features = myeloid_markers, group.by = "integrated_snn_res.1") + coord_flip()

```

```{r, echo=FALSE}

# myeloid immune cell type annotation
annotation_curated_myeloid <- read_excel("curated_annotation/curated_annotation_olalekan_integrated_Myeloid.xlsx")

Idents(Myeloid) <- Myeloid$integrated_snn_res.1
new_ids_main <- annotation_curated_myeloid$cell_subtype
names(new_ids_main) <- levels(Myeloid)
Myeloid <- RenameIdents(Myeloid, new_ids_main)
Myeloid@meta.data$cell_subtype <- Idents(Myeloid)

DimPlot(Myeloid, group.by = "cell_subtype", label = T)

```

```{r, echo=FALSE}

cell_counts <- FetchData(Myeloid, c("cell_subtype", "sample_id")) %>% filter(cell_subtype != "Contamination")

ggplot(cell_counts) +
  geom_bar(aes(x = sample_id, fill = cell_subtype), position = "fill")

ggplot(cell_counts) +
  geom_bar(aes(x = sample_id, fill = cell_subtype))

FeaturePlot(Myeloid, "IDO1", order = TRUE)

VlnPlot(Myeloid, "IDO1", pt.size = 1)
ggsave("output/olalekan_vlnplot_myeloid_ido1.pdf", height = 5, width = 7)

```

